### DISCUSSION TITLE
Added: LDM Notes for March 30, 2020

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#3323] [General]
## March 30, 2020

1. Record Monday

[C# Language Design Notes for March 30, 2020](https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-03-30.md)

**Comment by 333fred** [comment]
@CyrusNajmabadi I just fixed it.

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by YairHalberstadt** [comment]
> No separable value equality for non-records, right now. Default record equality is defined as field-based shallow equality.

Strongly agreed.

For me the most important aspects of records is that I can write something in one line as short as:

`record Request(int id, string url, data[] Data)` and have it just work. That I want inbuilt into the language, because **in scala I use it multiple times in almost every single file!**

I don't want records to be built of separable features because I would almost never use them. There's so many options once we step out of record land that it needs to be user customizable, and source generators seem the perfect way to do that.

**Comment by qrli** [comment]
> Recognizing the modreq and allowing access to init-only members, and calling any necessary "validator" on construction, are pretty simple features for VB. 

I have never seen the "validator" feature design. Is it mentioned anywhere?

**Comment by YairHalberstadt** [comment]
@qrli
https://github.com/jaredpar/csharplang/blob/record/proposals/recordsv3.md

**Comment by YairHalberstadt** [comment]
Also, @Agocke, thanks for putting up the meeting notes so quickly - it's really appreciated!

**Comment by Joe4evr** [comment]
Please consider allowing the use of `init` for non-readonly properties as well, since that would be the general purpose solution to #2328. I would really hate it if such a useful building block is only good for "Popsicle immutable" properties while cases where one works with data requiring a setter get left out in the cold *again*.

**Comment by orthoxerox** [comment]
> Default record equality is defined as field-based shallow equality.

And I think this is absolutely fine. People who need sequence equality for list members can either override `Equals` or write their own struct wrapper around `List<T>` that defines `Equals` as sequence equality.

**Comment by qrli** [comment]
Thanks for the link @YairHalberstadt 

After reading that, I feel the considered use cases for "validators" may be a bit less than I expected. Maybe it is the name, but to me it is not just "validators". It may also do:
- Precompute something and cache it in some mutable field
- Normalize the input
- Marshal data to be in a slightly different structure to be more efficient for read later

I'm not sure whether current "validator" design allows modifying init-only fields. If it allows, it could be fine.

**Comment by TonyValenti** [comment]
I definitely like the ability for “validators” to be able to precompute or modify input variables.

**Comment by CyrusNajmabadi** [comment]
In the last LDM we discussed not calling these 'validators' and instead just reusing the 'init' keyword in a manner like this:

```
init
{
}
```

This would basically just be code that runs after the constructor runs.  It would be able to do anyting that the constructor would normally be able to do (validate, compute, assign, cache, etc.).

IIRC, this was viewed highly positively.

**Comment by qrli** [comment]
That's cool.
Not right time for syntax, but `init() { ... }` would read more like a function and follows C# tradition than `init { ... }` 😀

**Comment by markm77** [comment]
> In the last LDM we discussed not calling these 'validators' and instead just reusing the 'init' keyword in a manner like this:
> 
> ```
> init
> {
> }
> ```
> 
> This would basically just be code that runs after the constructor runs. It would be able to do anyting that the constructor would normally be able to do (validate, compute, assign, cache, etc.).
> 
> IIRC, this was viewed highly positively.

How would this work with inheritance?

Could I use an object initialiser for a derived class that assigns properties to both base and derived classes? And could the validators then be automatically called in order, base class validator then derived class validator, to run any post-init checks/updates?

If so that would be really powerful.... in fact would such use of object initialisation and validators potentially remove the need for constructors and allow a very clean inheritance system without constructor chains?

**Comment by CyrusNajmabadi** [comment]
Similar to constructors, it would likely have to call into the base type to init.

> Could I use an object initialiser for a derived class that assigns properties to both base and derived classes?

Yes.

> If so that would be really powerful.... in fact would such use of object initialisation and validators potentially remove the need for constructors and allow a very clean inheritance system without constructor chains?

I don't see it being any cleaner/less-clean than constructors.  They're supposed to be effectively isomorphic.  Just one allows you to initialize directly after the object creation, one requires inside the creation.

**Comment by markm77** [comment]
> Similar to constructors, it would likely have to call into the base type to init.

I see. I was hoping _the compiler_ could automatically call the base class 'validator' before the derived class 'validator' as an improvement on how constructors work.

I'm hoping in future this records feature can extend to allowing multiple groups of properties to be easily combined/composed (i.e. simple multiple inheritance) which would be very convenient for building up data objects.

Wouldn't automating validator calls to the base class help keep things clean and reduce boilerplate as well as simplifying a future extension to support multiple inheritance?

**Comment by CyrusNajmabadi** [comment]
> I see. I was hoping the compiler could automatically call the base class 'validator' before the derived class 'validator' as an improvement on how constructors work.

yes.  that's what i meant :)

users would not control this.  it would be like a destructor.  the base destructor is getting called automatically :)

**Comment by markm77** [comment]
> users would not control this. it would be like a destructor. the base destructor is getting called automatically :)

Sounds great and a lot cleaner than having explicit ``` : base() ``` calls as is the case with constructors. That's the point I was trying to make - apologies if it wasn't clear.

If you go the route of property-level construction/destruction with optional post-constructors (validators) for whole object validation etc, would it also make sense to have optional pre-destructors for whole object tasks such as cleanup or logging the removal of an object?

These would be symmetric to the validators and called in the opposite order (derived class pre-destructor then base class pre-destructor).

I really like the idea of automatically called whole-object post-constructors/pre-destructors.... with the compiler handling calls up and down the inheritance chain....

**Comment by agocke** [comment]
> re-destructors for whole object tasks such as cleanup or logging the removal of an object

The only "destructor" in C# is finalizers. I don't think any change to how finalizers work is currently on the table. Personally, I find finalizers a pretty rough part of the language and would prefer we not modify that area any more than necessary.

**Comment by markm77** [comment]
> The only "destructor" in C# is finalizers. I don't think any change to how finalizers work is currently on the table. Personally, I find finalizers a pretty rough part of the language and would prefer we not modify that area any more than necessary.

Okay, I think I mis-understood something I read about records and property-level destruction (which I was curious about). But anyway as @CyrusNajmabadi mentioned, destructors/finalizers already are automatically called up the inheritance chain.

One final question (thanks for the illuminating discussion): I assume there will be a compiler warning if an object initializer omits a property that lacks an explicit default? (i.e. use of unspecified default values is assumed unintended). It seems to me use of C# language default values where not specified by the programmer can often conceal bugs and is not massively compatible with nullable reference types.

**Comment by CyrusNajmabadi** [comment]
> Sounds great and a lot cleaner than having explicit : base() calls as is the case with constructors. 

You don't need explicit `: base()` calls with constructors.  That will be given to you automatically if you emit it.

**Comment by markm77** [comment]
> You don't need explicit : base() calls with constructors. That will be given to you automatically if you emit it.

Thanks, apologies for my error.

**Comment by HaloFour** [comment]
I was playing with the proposed `modreq` approach for init-only members and I think I found a problem with it.  The issue is that if you apply `modreq` to the signature of the setter method, you have to also reference that `modreq` in the setter metadata of the property.  This causes all existing C# compilers to not just ignore the setter method, but the property as well.  It makes it illegal to even _read_ from the property directly.  You get the following error message when you attempt this:

> Error | CS1545 Property, indexer, or event 'TestClass.Property' is not supported by the language; try directly calling accessor methods 'TestClass.get_Property()' or 'TestClass.set_Property(?)'

As mentioned, the only way to read the property is to invoke the getter accessor directly, but this creates another problem as C# does not allow you to directly invoke accessor methods for properties that it can support.  So if someone were to work around this problem by calling `get_Property()` directly, upgrading to a supported version of the compiler would be a breaking change.  So would converting the init-only member to a normal setter, as the compiler would now recognize it as a property and disallow the direct accessor call.

I don't know if there is another way to apply the `modreq` that would avoid this.  I tried on both the return type and on the parameter and the behavior was the same.  In both cases I was required to also apply the `modreq` on the property metadata and C#/VS2019 behaved as described above.

Weirdly enough this behavior is not consistent across the ecosystem.  VB.NET has no problem reading from the property, for example.  And F# completely ignores the `modreq`, it has no problem reading from or writing to the property.  Actually, that seems like a fairly serious bug in F#.

**Comment by alrz** [comment]
>This would basically just be code that runs after the constructor runs. 

I would suggest the primary constructor on records replace the default ctor, so arguments would be accessible in the declared constructor.
```cs
data class Person(string name) {
  public Person() {
    if (name is null) throw new Exception();
  }
}
```
All the other constructors (with args) must either call the primary constructor or another constructor to make sure all fields are initialized.

**Comment by 333fred** [comment]
I can't agree with replacing the default constructor. I could see us allowing you to leave off the parens entirely to create a body, but validation will likely be done by whatever syntax we choose for validators, which will work for all user-defined types, not just records.

**Comment by markm77** [comment]
> One final question (thanks for the illuminating discussion): I assume there will be a compiler warning if an object initializer omits a property that lacks an explicit default? (i.e. use of unspecified default values is assumed unintended). It seems to me use of C# language default values where not specified by the programmer can often conceal bugs and is not massively compatible with nullable reference types.

The above isn't written very well but really my questions here should have been:
1. will there be a simple way to specify that certain properties must be initialized in the object initializer? (I think the answer here is that the only option will be value-checking in the future "post-constructor" which is not quite the same thing) 
2. will we be able to use e.g. non-nullable reference types for properties required to be initialised (those mentioned above) without having to initialize such properties with artificial default values e.g. in a constructor? Seems the answer is no from what I have read? Does this push us away from non-nullable types and NRT so we can allow required properties to start life as `null`? And then we check they are not `null` later in the "post-constructor"?

It would be really great if we could somehow specify parameters to be "required" in object initialization and then be able to use e.g. non-nullable reference types without artificial default values or compiler warnings....

Any thoughts?

**Comment by Joe4evr** [comment]
@markm77 No to both of your questions. Go see the discussion in #2328.
